### ISSUE TITLE
KNITRO not found on macOS

### ISSUE THREAD

**Issue by dpo** [jump-dev/KNITRO.jl#225]
I'm on macOS M1 and for some reason, KNITRO.jl isn't finding my installation of KNITRO:
```julia
(@v1.7) pkg> activate test-lin-cons/
  Activating project at `~/dev/JSO/NCL.jl/test-lin-cons`

(test-lin-cons) pkg> build KNITRO
    Building KNITRO → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/f880c76ceb1b5b776f4dd57b01c5b2d1954a56e6/build.log`
Precompiling project...
  2 dependencies successfully precompiled in 7 seconds (51 already precompiled)

julia> using KNITRO
[ Info: Precompiling KNITRO [67920dd8-b58e-52a8-8622-53c4cffbe346]

julia> KNITRO.has_knitro()
false

julia> ENV["KNITRODIR"]
"/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1"

julia> ENV["LD_LIBRARY_PATH"]
"/usr/local/lib:/usr/local/lib::/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib"

julia> ENV["DYLD_LIBRARY_PATH"]
"/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib:"

(test-lin-cons) pkg> st
      Status `~/dev/JSO/NCL.jl/test-lin-cons/Project.toml`
  [54578032] ADNLPModels v0.3.3
  [4076af6c] JuMP v1.1.1
  [67920dd8] KNITRO v0.13.0
  [792afdf1] NLPModelsJuMP v0.11.0
  [bec4dd0d] NLPModelsKnitro v0.5.7
```

I don't think `LD_LIBRARY_PATH` is particularly relevant on macOS. I used to have to set `DYLD_LIBRARY_PATH` instead, but that doesn't seem to do it either.

I can certainly use the `knitroampl` binary from the shell, so my license is ok. The `build.log` file is empty.

If I try to run the unit tests, I get messages of the form
```
Instantiation Knitro C interface: Error During Test at /Users/dpo/.julia/packages/KNITRO/6cfYO/test/C_wrapper.jl:37
  Got exception outside of a @test
  could not load library ""
  dlopen(.dylib, 0x0001): tried: '/.dylib' (no such file), '/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/.dylib' (no such file), '/Applications/Julia-1.7.app/Contents/Resources/julia/lib/julia/.dylib' (no such file), '/Applications/Julia-1.7.app/Contents/Resources/julia/bin/../lib/.dylib' (no such file), '.dylib' (no such file), '/usr/local/lib/.dylib' (no such file), '/usr/lib/.dylib' (no such file), '/.dylib' (no such file), '/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/.dylib' (no such file), '/Users/dpo/.julia/packages/KNITRO/6cfYO/test/.dylib' (no such file)
```

Any ideas?

**Comment by frapac** [comment]
That's indeed unexpected. `Pkg.build` looks for a shared library both in `$KNITRODIR/lib` and in `LD_LIBRARY_PATH`. Your KNITRODIR looks correct, so it should be able to find the library on your computer. What's the output of the following command? 
```shell
ls $KNITRODIR/lib
```

**Comment by dpo** [comment]
```
$ ls -l $KNITRODIR/lib
total 25200
lrwxr-xr-x@ 1 dpo  staff    15B Feb 10 13:29 libknitro.a -> libknitro1301.a
lrwxr-xr-x@ 1 dpo  staff    19B Feb 10 13:29 libknitro.dylib -> libknitro1301.dylib
lrwxr-xr-x@ 1 dpo  staff    19B Feb 10 13:29 libknitro.dylib.13 -> libknitro1301.dylib
lrwxr-xr-x@ 1 dpo  staff    19B Feb 10 13:29 libknitro.dylib.13.0 -> libknitro1301.dylib
lrwxr-xr-x@ 1 dpo  staff    19B Feb 10 13:29 libknitro.dylib.13.0.1 -> libknitro1301.dylib
-rw-r--r--@ 1 dpo  staff   6.6M Feb 10 13:25 libknitro1301.a
-rwxr-xr-x@ 1 dpo  staff   5.1M Feb 10 13:25 libknitro1301.dylib
-rw-r--r--@ 1 dpo  staff   647K Feb  1 13:26 libomp.dylib
```

**Comment by frapac** [comment]
ok. And what's the output of the following command in Julia?
```julia
using Libdl
knitro_lib = joinpath(ENV["KNITRODIR"], "lib", "libknitro.dylib")
Libdl.dlopen_e(knitro_lib)

```

**Comment by dpo** [comment]
```
julia> using Libdl

julia> knitro_lib = joinpath(ENV["KNITRODIR"], "lib", "libknitro.dylib")
"/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/libknitro.dylib"

julia> Libdl.dlopen_e(knitro_lib)
Ptr{Nothing} @0x0000000000000000

julia> Libdl.dlopen(knitro_lib)
ERROR: could not load library "/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/libknitro.dylib"
dlopen(/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/libknitro.dylib, 0x0001): tried: '/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/libknitro.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/libknitro.dylib' (no such file), '/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/libknitro.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/libknitro1301.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/libknitro1301.dylib' (no such file), '/Users/dpo/local/solvers/knitro/knitro-13.0.1-MacOS-M1/lib/libknitro1301.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))
Stacktrace:
 [1] dlopen(s::String, flags::UInt32; throw_error::Bool)
   @ Base.Libc.Libdl ./libdl.jl:117
 [2] dlopen (repeats 2 times)
   @ ./libdl.jl:117 [inlined]
 [3] top-level scope
   @ REPL[4]:1
```

Is this due to the fact I have to run Julia for x86_64 on the M1, yet the Knitro lib is for M1? I just tried with Ipopt and it seems to be working... though the unit tests hang midway... 🤔

**Comment by dpo** [comment]
Would the Knitro 86_64 run on the M1? Is it possible to install it alongside the M1 lib?

**Comment by frapac** [comment]
I think you can download Knitro x86_64 on the M1. Not sure it will work out of the box (the x86_64 build is compiled with MKL). Have you tried running Knitro with rosetta?

**Comment by odow** [comment]
Is this resolved? You should be able to run KNITRO x86_64 on the M1 with Julia x86_64, Or KNITRO arm64 with Julia arm64. 

I have an M1, but not a KNITRO license, so I can't test unfortunately.

x-ref: https://github.com/jump-dev/JuMP.jl/issues/2971. If you previously tried on Julia 1.7, try with the latest 1.8-rc4. It fixes a lot of issues.

**Comment by dpo** [comment]
I'm still getting the same errors as above after installing KNITRO 13.1 for x86_64 and running it in Julia 1.7.3 for x86_64. I'll test Julia 1.8-rc4 as soon as it appears on juliaup. I can run `knitroampl` just fine from the command line.

**Comment by dpo** [comment]
Upgrading juliaup gave access to 1.8-rc4. The results are more promising but a bunch of tests fail. `KNITRO.has_knitro()` finally returns `true`. The complete output is here: https://gist.github.com/dpo/2df7b64b32855619206b608517050413

**Comment by odow** [comment]
The test failures are #224 and #228, but it looks like the M1 is working fine.

**Comment by odow** [comment]
Closing because tests are passing (although we need to fix a bunch of excluded ones): #230
